### PR TITLE
b2c fixes, refactoring, addressing pr comments

### PR DESCRIPTION
**Pull request by SomkaPe**
support of B2C:
- In case of absence of required claims in b2C idToken skip storing tokens in the cache
- cache look up based on environments aliases and tenant id instead of authority
- refactoring
- addressing some pr comments

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 1]
```diff
@@ -234,10 +234,9 @@ private void SaveTokenResponseToCache()
                 AuthenticationRequestParameters.RequestContext.Logger.Info(msg);
                 AuthenticationRequestParameters.RequestContext.Logger.InfoPii(msg);
 
-                // todo should we return idToken from SaveAccessAndRefreshToken as well ?
-                // problem - if AT is not stored we will fail  ?
-                MsalAccessTokenItem = TokenCache.SaveAccessAndRefreshToken(AuthenticationRequestParameters, Response);
-                MsalIdTokenItem = TokenCache.GetIdTokenCacheItem(MsalAccessTokenItem.GetIdTokenItemKey(), AuthenticationRequestParameters.RequestContext);
+                var tuple = TokenCache.SaveAccessAndRefreshToken(AuthenticationRequestParameters, Response);
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:237]: Have a look at c# 7 tuples - there's a nicer way of supporting tuples. 

https://blogs.msdn.microsoft.com/mazhou/2017/05/26/c-7-series-part-1-value-tuples/

**Review by bgavrilMS — APPROVED** [review]

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 2]
```diff
@@ -239,37 +200,107 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
...
+                requestParams.RequestContext.Logger.InfoPii(msg);
+            }
+
+            foreach (var cacheItem in accessTokenItemList)
+            {
+                tokenCacheAccessor.DeleteAccessToken(cacheItem.GetKey(), requestParams.RequestContext);
+            }
+        }
+
+        private bool ContainsClaimsRequiredToStoreInCahce(IdToken idToken, RequestContext requestContext)
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >Cahce [](start = 52, length = 5)

spelling: Cache

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 3]
```diff
@@ -239,37 +200,107 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
...
+        {
+            if (idToken == null || string.IsNullOrEmpty(idToken.TenantId) || string.IsNullOrEmpty(idToken.PreferredUsername))
+            {
+                string msg;
+                if (idToken == null)
+                {
+                    msg = "Skipping storing tokens in the cahe - no IdToken in token response";
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >cahe [](start = 58, length = 4)

spelling

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 4]
```diff
@@ -239,37 +200,107 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
...
+                }
+                else
+                {
+                    msg = "Skipping storing tokens in the cahe - IdToken does not contain required claims";
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >cahe [](start = 58, length = 4)

spelling

File: msal/src/Microsoft.Identity.Client/TokenCache.cs [hunk 5]
```diff
@@ -112,123 +112,84 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
             BeforeWrite?.Invoke(args);
         }
 
-        internal MsalAccessTokenCacheItem SaveAccessAndRefreshToken(AuthenticationRequestParameters requestParams,
-            MsalTokenResponse response)
+        internal Tuple<MsalAccessTokenCacheItem, MsalIdTokenCacheItem> SaveAccessAndRefreshToken
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs:115]: >Tuple<MsalAccessTokenCacheItem, MsalIdTokenCacheItem> [](start = 17, length = 53)

nit: I believe Bogdan was saying the same. You can use a value tuple instead

**Comment by henrik-me** [comment]
nit: readability will increase for me if code can be in one line. Seems like all lines are cut to something like 80 chars... it makes it harder for me to read the code.

**Review by henrik-me — APPROVED** [review]
:shipit:

**Comment by henrik-me** [comment]
few small "issues" when fixed this is ready to get into the repo

File: msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs (hunk 1 above)
Comment (SomkaPe) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:237]: I have tried to use it but looks like I have to add some nuget packages to support it so I gave up )

**Pull request merged**